File: Models/OCRParser.swift
import Foundation

/// Parses raw OCR text from a recipe photo into structured recipe fields.
/// Pure Swift — no Apple frameworks.
///
/// The OCR text is expected to be a full recipe with a title, ingredients
/// list, and instructions. This parser uses heuristic line-by-line analysis
/// to extract structure from flat text.

struct ParsedRecipe: Codable, Equatable {
    var title: String
    var ingredients: [ParsedIngredient]
    var instructions: [String]
    var servings: Int?
    var prepTimeMinutes: Int?
    var cookTimeMinutes: Int?
}

struct ParsedIngredient: Codable, Equatable {
    var name: String
    var quantity: Double
    var unit: String

    init(name: String, quantity: Double = 1, unit: String = "") {
        self.name = name
        self.quantity = quantity
        self.unit = unit
    }
}

/// Section markers that indicate the start of an ingredients list.
private let ingredientHeaders = [
    "ingredients", "ingredient list", "what you need",
    "you will need", "you'll need", "shopping list",
]

/// Section markers that indicate the start of instructions.
private let instructionHeaders = [
    "instructions", "directions", "method", "steps",
    "preparation", "how to make", "procedure",
]

/// Parses raw OCR text into a `ParsedRecipe`.
func parseRecipeText(_ text: String) -> ParsedRecipe {
    let lines = text.components(separatedBy: .newlines)
        .map { $0.trimmingCharacters(in: .whitespaces) }

    var title = ""
    var ingredients: [ParsedIngredient] = []
    var instructions: [String] = []
    var servings: Int? = nil
    var prepTime: Int? = nil
    var cookTime: Int? = nil

    enum Section { case header, ingredients, instructions, unknown }
    var currentSection: Section = .header

    for line in lines {
        let lower = line.lowercased()

        // Skip blank lines
        if line.isEmpty { continue }

        // Detect section headers
        if ingredientHeaders.contains(where: { lower.hasPrefix($0) }) {
            currentSection = .ingredients
            continue
        }
        if instructionHeaders.contains(where: { lower.hasPrefix($0) }) {
            currentSection = .instructions
            continue
        }

        // Extract metadata from any section
        if let s = parseServings(lower) {
            servings = s
            continue
        }
        if let t = parseTimeField(lower, prefixes: ["prep time", "prep"]) {
            prepTime = t
            continue
        }
        if let t = parseTimeField(lower, prefixes: ["cook time", "cooking time", "cook"]) {
            cookTime = t
            continue
        }
        if let t = parseTimeField(lower, prefixes: ["total time", "total"]) {
            // If we have prep but not cook, derive cook from total
            if prepTime != nil && cookTime == nil {
                cookTime = max(0, t - (prepTime ?? 0))
            } else if prepTime == nil && cookTime == nil {
                // Just store as cook time
                cookTime = t
            }
            continue
        }

        switch currentSection {
        case .header:
            // First non-metadata, non-blank line is the title
            if title.isEmpty {
                title = cleanRecipeTitle(line)
            }
        case .ingredients:
            if let ingredient = parseIngredientLine(line) {
                ingredients.append(ingredient)
            }
        case .instructions:
            let cleaned = cleanInstructionLine(line)
            if !cleaned.isEmpty {
                instructions.append(cleaned)
            }
        case .unknown:
            break
        }
    }

    // Fallback: if no section headers were found, try parsing all non-title
    // lines as ingredients. Real-world recipe photos often lack headers.
    if ingredients.isEmpty && instructions.isEmpty {
        for line in lines {
            if line.isEmpty { continue }
            let lower = line.lowercased()
            // Skip the title line
            if line == title || line == cleanRecipeTitle(line) && line == title { continue }
            // Skip metadata lines
            if parseServings(lower) != nil { continue }
            if parseTimeField(lower, prefixes: ["prep time", "prep"]) != nil { continue }
            if parseTimeField(lower, prefixes: ["cook time", "cooking time", "cook"]) != nil { continue }
            if parseTimeField(lower, prefixes: ["total time", "total"]) != nil { continue }
            // Try as ingredient
            if let ingredient = parseIngredientLine(line) {
                ingredients.append(ingredient)
            }
        }
    }

    return ParsedRecipe(
        title: title,
        ingredients: ingredients,
        instructions: instructions,
        servings: servings,
        prepTimeMinutes: prepTime,
        cookTimeMinutes: cookTime
    )
}

/// Parses an ingredient line like "2 cups flour" or "1/2 lb chicken breast"
func parseIngredientLine(_ line: String) -> ParsedIngredient? {
    var cleaned = line.trimmingCharacters(in: .whitespaces)
    guard !cleaned.isEmpty else { return nil }

    // Strip list markers
    if cleaned.hasPrefix("- ") || cleaned.hasPrefix("• ") || cleaned.hasPrefix("* ") {
        cleaned = String(cleaned.dropFirst(2))
    }
    cleaned = cleaned.trimmingCharacters(in: .whitespaces)
    guard !cleaned.isEmpty else { return nil }

    // Reuse the list line parser logic
    if let parsed = parseListLine(cleaned) {
        return ParsedIngredient(name: parsed.name, quantity: parsed.quantity, unit: parsed.unit)
    }
    return ParsedIngredient(name: cleaned)
}

/// Extracts servings from a line like "Serves 4" or "Servings: 6"
func parseServings(_ lower: String) -> Int? {
    let patterns = ["serves", "servings:", "servings", "yield:", "yield", "makes"]
    for prefix in patterns {
        if lower.hasPrefix(prefix) {
            let rest = lower.dropFirst(prefix.count)
                .trimmingCharacters(in: .whitespaces)
                .trimmingCharacters(in: CharacterSet(charactersIn: ":"))
                .trimmingCharacters(in: .whitespaces)
            // Extract first number
            let digits = rest.prefix(while: { $0.isNumber })
            if let n = Int(digits), n > 0 { return n }
        }
    }
    return nil
}

/// Extracts time in minutes from a line like "Prep time: 20 min"
func parseTimeField(_ lower: String, prefixes: [String]) -> Int? {
    for prefix in prefixes {
        if lower.hasPrefix(prefix) {
            let rest = lower.dropFirst(prefix.count)
                .trimmingCharacters(in: .whitespaces)
                .trimmingCharacters(in: CharacterSet(charactersIn: ":"))
                .trimmingCharacters(in: .whitespaces)
            return parseTimeString(String(rest))
        }
    }
    return nil
}

/// Parses time strings like "20 min", "1 hour", "1h 30m", "90 minutes"
func parseTimeString(_ text: String) -> Int? {
    let lower = text.lowercased()
    var totalMinutes = 0
    var foundAny = false

    // Match hours: "1 hour", "2 hours", "1hr", "1h"
    // Use word-boundary-safe pattern: digits + optional space + hour word
    if let range = lower.range(
        of: "(\\d+)\\s*(?:hours?|hrs?|h)(?:[^a-z]|$)",
        options: .regularExpression
    ) {
        let match = String(lower[range])
        let digits = match.prefix(while: { $0.isNumber })
        if let h = Int(digits) {
            totalMinutes += h * 60
            foundAny = true
        }
    }

    // Match minutes: "30 minutes", "20 min", "30m"
    // Must not match "m" inside "hour" — require digits immediately before
    if let range = lower.range(
        of: "(\\d+)\\s*(?:minutes?|mins?|m)(?:[^a-z]|$)",
        options: .regularExpression
    ) {
        let match = String(lower[range])
        let digits = match.prefix(while: { $0.isNumber })
        if let m = Int(digits) {
            totalMinutes += m
            foundAny = true
        }
    }

    // Bare number — assume minutes
    if !foundAny {
        let digits = lower.prefix(while: { $0.isNumber })
        if let m = Int(digits), m > 0 {
            return m
        }
    }

    return foundAny ? totalMinutes : nil
}

/// Cleans a recipe title (removes trailing colons, excess whitespace).
func cleanRecipeTitle(_ title: String) -> String {
    var cleaned = title.trimmingCharacters(in: .whitespaces)
    if cleaned.hasSuffix(":") {
        cleaned = String(cleaned.dropLast()).trimmingCharacters(in: .whitespaces)
    }
    return cleaned
}

/// Cleans an instruction line (removes numbered prefixes like "1. ", "Step 2: ").
func cleanInstructionLine(_ line: String) -> String {
    var cleaned = line.trimmingCharacters(in: .whitespaces)

    // Remove "Step N:" or "Step N." prefix
    if cleaned.lowercased().hasPrefix("step") {
        let rest = cleaned.dropFirst(4).trimmingCharacters(in: .whitespaces)
        // Drop the number and separator
        let afterNum = rest.drop(while: { $0.isNumber })
            .trimmingCharacters(in: .whitespaces)
        if afterNum.hasPrefix(":") || afterNum.hasPrefix(".") || afterNum.hasPrefix(")") {
            cleaned = String(afterNum.dropFirst()).trimmingCharacters(in: .whitespaces)
        }
    }
    // Remove "N. " or "N) " prefix
    else if cleaned.first?.isNumber == true {
        let afterNum = cleaned.drop(while: { $0.isNumber })
            .trimmingCharacters(in: .whitespaces)
        if afterNum.hasPrefix(".") || afterNum.hasPrefix(")") {
            cleaned = String(afterNum.dropFirst()).trimmingCharacters(in: .whitespaces)
        }
    }

    return cleaned
}

File: recipe-app-rn/src/lib/ocrParser.ts
/**
 * Parses raw OCR text from a recipe photo into structured recipe fields using
 * heuristic line-by-line analysis (title, ingredients, instructions, servings,
 * prep/cook time). Reuses the shopping-list line parser for ingredient lines.
 * 1:1 port of `SharedLogic/OCRParser.swift` (framework-free).
 */

import { parseListLine, ParsedListItem } from './listLineParser';

export type ParsedIngredient = { name: string; quantity: number; unit: string };

export type ParsedRecipe = {
  title: string;
  ingredients: ParsedIngredient[];
  instructions: string[];
  servings: number | null;
  prepTimeMinutes: number | null;
  cookTimeMinutes: number | null;
};

/** Section markers that indicate the start of an ingredients list. */
const ingredientHeaders = [
  'ingredients', 'ingredient list', 'what you need',
  'you will need', "you'll need", 'shopping list',
];

/** Section markers that indicate the start of instructions. */
const instructionHeaders = [
  'instructions', 'directions', 'method', 'steps',
  'preparation', 'how to make', 'procedure',
];

/** True if `ch` is an ASCII digit (mirrors Swift `Character.isNumber` for our inputs). */
function isDigit(ch: string): boolean {
  return ch >= '0' && ch <= '9';
}

/** Leading run of digits (mirrors Swift `prefix(while: { $0.isNumber })`). */
function leadingDigits(s: string): string {
  let i = 0;
  while (i < s.length && isDigit(s[i])) i += 1;
  return s.slice(0, i);
}

/** Drops the leading run of digits (mirrors Swift `drop(while: { $0.isNumber })`). */
function dropLeadingDigits(s: string): string {
  let i = 0;
  while (i < s.length && isDigit(s[i])) i += 1;
  return s.slice(i);
}

/** Trims whitespace off both ends (mirrors Swift `trimmingCharacters(in: .whitespaces)`). */
function trimWhitespace(s: string): string {
  return s.replace(/^[ \t]+/, '').replace(/[ \t]+$/, '');
}

/** Trims any of `chars` off both ends (mirrors Swift `trimmingCharacters(in:)`). */
function trimCharacters(s: string, chars: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && chars.includes(s[start])) start += 1;
  while (end > start && chars.includes(s[end - 1])) end -= 1;
  return s.slice(start, end);
}

/** Strict integer parse mirroring Swift's `Int(String)` (rejects empty/garbage). */
function parseInt10(s: string): number | null {
  if (s.length === 0) return null;
  if (!/^[+-]?\d+$/.test(s)) return null;
  const n = Number(s);
  if (Number.isNaN(n)) return null;
  return n;
}

/** Parses raw OCR text into a `ParsedRecipe`. */
export function parseRecipeText(text: string): ParsedRecipe {
  const lines = text.split(/\r\n|\n|\r/).map((l) => trimWhitespace(l));

  let title = '';
  const ingredients: ParsedIngredient[] = [];
  const instructions: string[] = [];
  let servings: number | null = null;
  let prepTime: number | null = null;
  let cookTime: number | null = null;

  type Section = 'header' | 'ingredients' | 'instructions' | 'unknown';
  let currentSection: Section = 'header';

  for (const line of lines) {
    const lower = line.toLowerCase();

    // Skip blank lines
    if (line.length === 0) continue;

    // Detect section headers
    if (ingredientHeaders.some((h) => lower.startsWith(h))) {
      currentSection = 'ingredients';
      continue;
    }
    if (instructionHeaders.some((h) => lower.startsWith(h))) {
      currentSection = 'instructions';
      continue;
    }

    // Extract metadata from any section
    const s = parseServings(lower);
    if (s !== null) {
      servings = s;
      continue;
    }
    const prep = parseTimeField(lower, ['prep time', 'prep']);
    if (prep !== null) {
      prepTime = prep;
      continue;
    }
    const cook = parseTimeField(lower, ['cook time', 'cooking time', 'cook']);
    if (cook !== null) {
      cookTime = cook;
      continue;
    }
    const total = parseTimeField(lower, ['total time', 'total']);
    if (total !== null) {
      // If we have prep but not cook, derive cook from total
      if (prepTime !== null && cookTime === null) {
        cookTime = Math.max(0, total - (prepTime ?? 0));
      } else if (prepTime === null && cookTime === null) {
        // Just store as cook time
        cookTime = total;
      }
      continue;
    }

    switch (currentSection) {
      case 'header':
        // First non-metadata, non-blank line is the title
        if (title.length === 0) {
          title = cleanRecipeTitle(line);
        }
        break;
      case 'ingredients': {
        const ingredient = parseIngredientLine(line);
        if (ingredient !== null) {
          ingredients.push(ingredient);
        }
        break;
      }
      case 'instructions': {
        const cleaned = cleanInstructionLine(line);
        if (cleaned.length > 0) {
          instructions.push(cleaned);
        }
        break;
      }
      default: // 'unknown'
        break;
    }
  }

  // Fallback: if no section headers were found, try parsing all non-title
  // lines as ingredients. Real-world recipe photos often lack headers.
  if (ingredients.length === 0 && instructions.length === 0) {
    for (const line of lines) {
      if (line.length === 0) continue;
      const lower = line.toLowerCase();
      // Skip the title line
      if (line === title || (line === cleanRecipeTitle(line) && line === title)) continue;
      // Skip metadata lines
      if (parseServings(lower) !== null) continue;
      if (parseTimeField(lower, ['prep time', 'prep']) !== null) continue;
      if (parseTimeField(lower, ['cook time', 'cooking time', 'cook']) !== null) continue;
      if (parseTimeField(lower, ['total time', 'total']) !== null) continue;
      // Try as ingredient
      const ingredient = parseIngredientLine(line);
      if (ingredient !== null) {
        ingredients.push(ingredient);
      }
    }
  }

  return {
    title,
    ingredients,
    instructions,
    servings,
    prepTimeMinutes: prepTime,
    cookTimeMinutes: cookTime,
  };
}

/** Parses an ingredient line like "2 cups flour" or "1/2 lb chicken breast". */
export function parseIngredientLine(line: string): ParsedIngredient | null {
  let cleaned = trimWhitespace(line);
  if (cleaned.length === 0) return null;

  // Strip list markers
  if (cleaned.startsWith('- ') || cleaned.startsWith('• ') || cleaned.startsWith('* ')) {
    cleaned = cleaned.slice(2);
  }
  cleaned = trimWhitespace(cleaned);
  if (cleaned.length === 0) return null;

  // Reuse the list line parser logic
  const parsed: ParsedListItem | null = parseListLine(cleaned);
  if (parsed !== null) {
    return { name: parsed.name, quantity: parsed.quantity, unit: parsed.unit };
  }
  return { name: cleaned, quantity: 1, unit: '' };
}

/** Extracts servings from a line like "Serves 4" or "Servings: 6". */
export function parseServings(lower: string): number | null {
  const patterns = ['serves', 'servings:', 'servings', 'yield:', 'yield', 'makes'];
  for (const prefix of patterns) {
    if (lower.startsWith(prefix)) {
      let rest = trimWhitespace(lower.slice(prefix.length));
      rest = trimCharacters(rest, ':');
      rest = trimWhitespace(rest);
      // Extract first number
      const digits = leadingDigits(rest);
      const n = parseInt10(digits);
      if (n !== null && n > 0) return n;
    }
  }
  return null;
}

/** Extracts time in minutes from a line like "Prep time: 20 min". */
export function parseTimeField(lower: string, prefixes: string[]): number | null {
  for (const prefix of prefixes) {
    if (lower.startsWith(prefix)) {
      let rest = trimWhitespace(lower.slice(prefix.length));
      rest = trimCharacters(rest, ':');
      rest = trimWhitespace(rest);
      return parseTimeString(rest);
    }
  }
  return null;
}

/** Parses time strings like "20 min", "1 hour", "1h 30m", "90 minutes". */
export function parseTimeString(text: string): number | null {
  const lower = text.toLowerCase();
  let totalMinutes = 0;
  let foundAny = false;

  // Match hours: "1 hour", "2 hours", "1hr", "1h"
  const hourMatch = /(\d+)\s*(?:hours?|hrs?|h)(?:[^a-z]|$)/.exec(lower);
  if (hourMatch !== null) {
    const match = hourMatch[0];
    const digits = leadingDigits(match);
    const h = parseInt10(digits);
    if (h !== null) {
      totalMinutes += h * 60;
      foundAny = true;
    }
  }

  // Match minutes: "30 minutes", "20 min", "30m"
  const minMatch = /(\d+)\s*(?:minutes?|mins?|m)(?:[^a-z]|$)/.exec(lower);
  if (minMatch !== null) {
    const match = minMatch[0];
    const digits = leadingDigits(match);
    const m = parseInt10(digits);
    if (m !== null) {
      totalMinutes += m;
      foundAny = true;
    }
  }

  // Bare number — assume minutes
  if (!foundAny) {
    const digits = leadingDigits(lower);
    const m = parseInt10(digits);
    if (m !== null && m > 0) {
      return m;
    }
  }

  return foundAny ? totalMinutes : null;
}

/** Cleans a recipe title (removes trailing colons, excess whitespace). */
export function cleanRecipeTitle(title: string): string {
  let cleaned = trimWhitespace(title);
  if (cleaned.endsWith(':')) {
    cleaned = trimWhitespace(cleaned.slice(0, -1));
  }
  return cleaned;
}

/** Cleans an instruction line (removes numbered prefixes like "1. ", "Step 2: "). */
export function cleanInstructionLine(line: string): string {
  let cleaned = trimWhitespace(line);

  // Remove "Step N:" or "Step N." prefix
  if (cleaned.toLowerCase().startsWith('step')) {
    const rest = trimWhitespace(cleaned.slice(4));
    // Drop the number and separator
    const afterNum = trimWhitespace(dropLeadingDigits(rest));
    if (afterNum.startsWith(':') || afterNum.startsWith('.') || afterNum.startsWith(')')) {
      cleaned = trimWhitespace(afterNum.slice(1));
    }
  } else if (cleaned.length > 0 && isDigit(cleaned[0])) {
    // Remove "N. " or "N) " prefix
    const afterNum = trimWhitespace(dropLeadingDigits(cleaned));
    if (afterNum.startsWith('.') || afterNum.startsWith(')')) {
      cleaned = trimWhitespace(afterNum.slice(1));
    }
  }

  return cleaned;
}
